### PR TITLE
Variables cleanup

### PR DESCRIPTION
**Pull request by webrouse**
Part of: https://keboola.atlassian.net/browse/COM-1137

### CODE REVIEW

File: internal/pkg/manifest/manifest_test.go [hunk 1]
```diff
@@ -667,10 +663,8 @@ func cyclicDependencyJson() string {
       "path": "variables",
       "relations": [
         {
-          "target": {
-            "componentId": "keboola.variables",
-            "id": "222"
-          },
+          "componentId": "keboola.variables",
```
Comment (webrouse) [inline internal/pkg/manifest/manifest_test.go:666]: Jednoduchsi zapis `variablesFor` relacie, bez `target`.

File: test/pull-variables-add-relation/out/.keboola/manifest.json [hunk 2]
```diff
@@ -53,10 +51,6 @@
           "path": "values/default",
           "relations": [
             {
-              "target": {
-                "componentId": "ex-generic-v2",
-                "id": "%%TEST_BRANCH_MAIN_CONFIG_WITH_VARIABLES_1_ID%%"
-              },
               "type": "variablesValuesFor"
```
Comment (webrouse) [inline test/pull-variables-add-relation/out/.keboola/manifest.json:54]: Pri `variablesValuesFor` relacii vyhodene vsetky IDs, pretoze `values` samostatne nedavaju zmysel, a musia mat rovnaky `target` ako `variablesFor`. Tj. tato relacia iba oznacuje, ktory z viacerych `rows` su `default values`, ... samotne napojenie na druhy config, je vo `variables` konfiguracii, pomocou `variablesFor` relacie.

Vid cely subor: 
https://github.com/keboola/keboola-as-code/blob/d6fc4897eb08e63b1699d69efb914ac4b2dd9f1f/test/pull-variables-add-relation/out/.keboola/manifest.json#L36-L59

Kedze v `manifest.json` nie su uz duplicitne IDs, tak odpadavaju nejake edge-cases, ak by sa napr. manualne zadalo rozne ID.

File: internal/pkg/model/relation.go [hunk 3]
```diff
@@ -41,13 +42,12 @@ func (t RelationType) Type() RelationType {
...
-	OtherSideKey(owner Key) Key               // get key of the other side
-	IsDefinedInManifest() bool                // if true, relation will be present in the manifest
-	IsDefinedInApi() bool                     // if true, relation will be present in API calls
-	NewOtherSideRelation(owner Key) Relation  // create the new other side relation, for example VariablesFor -> VariablesFrom
+	Desc() string                                 // human-readable description
+	Key() string                                  // unique key within the object on which the relation is defined, for sorting and comparing
+	ParentKey(relationDefinedOn Key) (Key, error) // if relation type is parent <-> child, then parent key is returned, otherwise nil
+	IsDefinedInManifest() bool                    // if true, relation will be present in the manifest
+	IsDefinedInApi() bool                         // if true, relation will be present in API calls
+	NewOtherSideRelation(relationDefinedOn Object, allObjects *StateObjects) (otherSide Key, relation Relation, err error)
```
Comment (webrouse) [inline internal/pkg/model/relation.go:50]: Namiesto `OtherSideKey`, a `NewOtherSideRelation`, je uz iba metoda `NewOtherSideRelation`:
- obe hodnoty spolu suvisia, `key` hovori, ktory objekt je na druhej strane, a `relation` aka je relation na druhej strane, definovana na tomto objekte
- metoda dostava naviac `allObjects`, ak relacia potrebuje aj info o inych objektoch, ... napr. aky je `parent`, ... vyuziva sa to pri` variablesValuesFor`, ... aby nemuselo byt ulozene duplicitne ID v `manifest.json` (vid ostatne komentare).

File: internal/pkg/model/relation.go [hunk 4]
```diff
@@ -159,6 +159,21 @@ func (v Relations) GetByType(t RelationType) Relations {
 	return out
 }
 
+func (v Relations) GetOneByType(t RelationType) (Relation, error) {
```
Comment (webrouse) [inline internal/pkg/model/relation.go:162]: Presunute do metody, pouziva sa na viacerych miestach.

File: internal/pkg/diff/diff.go [hunk 5]
```diff
@@ -131,7 +131,8 @@ func (d *Differ) diffState(state model.ObjectState) (*Result, error) {
 	// Diff
 	for _, field := range diffFields {
 		reporter := d.diffValues(
-			state.Key(),
```
Comment (webrouse) [inline internal/pkg/diff/diff.go:134]: `diff.Reporter` obsahuje namiesto `Key` oba, `remote/local` objekty, ktore porovnava.

File: internal/pkg/diff/reporter.go [hunk 6]
```diff
@@ -95,14 +97,16 @@ func (r *Reporter) isPathHidden() bool {
...
-func (r *Reporter) relationToString(relation model.Relation) string {
+func (r *Reporter) relationToString(relation model.Relation, definedOn model.Object, allObjects *model.StateObjects) string {
 	otherSideDesc := ``
-	otherSideKey := relation.OtherSideKey(r.objectKey)
-	if otherSide, found := r.state.Get(otherSideKey); found {
-		otherSideDesc = `"` + otherSide.Path() + `"`
-	}
-	if len(otherSideDesc) == 0 {
-		otherSideDesc = otherSideKey.Desc()
+	otherSideKey, _, err := relation.NewOtherSideRelation(definedOn, allObjects)
```
Comment (webrouse) [inline internal/pkg/diff/reporter.go:102]: `OtherSideKey` -> `NewOtherSideRelation`

File: internal/pkg/diff/reporter.go [hunk 7]
```diff
@@ -95,14 +97,16 @@ func (r *Reporter) isPathHidden() bool {
...
+	if err == nil {
```
Comment (webrouse) [inline internal/pkg/diff/reporter.go:103]: Pridana podmienka, kedze `NewOtherSideRelation` vracia `error`.

File: internal/pkg/manifest/manifest.go [hunk 8]
```diff
@@ -94,6 +94,11 @@ func LoadManifest(fs filesystem.Fs) (*Manifest, error) {
 		}
 	}
 
+	// Validate
```
Comment (webrouse) [inline internal/pkg/manifest/manifest.go:97]: Validacia `manifest.json` presunuta pred `Connect records together and resolve parent path`.
Ak je manifest nevalidny, nema zmysel pokracovat -> a chybove hlasky su viac zrozumitelne, ako keby to spadlo az dalej.

File: internal/pkg/mapper/relations/load.go [hunk 9]
```diff
@@ -27,91 +27,91 @@ func (m *relationsMapper) MapAfterLocalLoad(recipe *model.LocalLoadRecipe) error
...
 	errors := utils.NewMultiError()
 
-	// Link
+	// Link and validate
 	for _, object := range event.NewObjects {
 		if o, ok := object.(model.ObjectWithRelations); ok {
 			if err := m.linkRelations(o, event); err != nil {
 				errors.Append(err)
 			}
+			if err := m.validateRelations(o); err != nil {
```
Comment (webrouse) [inline internal/pkg/mapper/relations/load.go:36]: Po `linkRelations` sa hned vola `validateRelations` nad tym istym objektom.
Chybove hlasky k jednemu objektu su tak za sebou, a vystup je prehladnejsi.

File: internal/pkg/mapper/relations/load.go [hunk 10]
```diff
@@ -27,91 +27,91 @@ func (m *relationsMapper) MapAfterLocalLoad(recipe *model.LocalLoadRecipe) error
...
 	}
 
-	// Validate
-	for _, object := range event.NewObjects {
-		if o, ok := object.(model.ObjectWithRelations); ok {
-			m.validateRelations(o)
-		}
+	// Log errors as warning
+	if errors.Len() > 0 {
+		m.Logger.Warn(utils.PrefixError(`Warning`, errors))
```
Comment (webrouse) [inline internal/pkg/mapper/relations/load.go:44]: Vsetky chyby v `relations` sa loguju iba ako `watning`, ... chybne relacie sa ignoruju, vid `relations.Remove(relation)` v dalsom kode.

File: internal/pkg/mapper/relations/load.go [hunk 11]
```diff
@@ -27,91 +27,91 @@ func (m *relationsMapper) MapAfterLocalLoad(recipe *model.LocalLoadRecipe) error
...
 	errors := utils.NewMultiError()
-	for _, relation := range object.GetRelations() {
-		// Get relation other side
-		thisSideKey := object.Key()
-		otherSideKey := relation.OtherSideKey(thisSideKey)
+	relations := object.GetRelations()
+
+	for _, relation := range relations {
+		// Get other side relation
+		otherSideKey, otherSideRelation, err := relation.NewOtherSideRelation(object, event.AllObjects)
```
Comment (webrouse) [inline internal/pkg/mapper/relations/load.go:57]: Namiesto `OtherSideKey` a `NewOtherSideRelation` -> uz iba `NewOtherSideRelation`.

File: internal/pkg/mapper/relations/load.go [hunk 12]
```diff
@@ -27,91 +27,91 @@ func (m *relationsMapper) MapAfterLocalLoad(recipe *model.LocalLoadRecipe) error
...
-			err := fmt.Errorf(
-				`%s not found, referenced from %s, by relation "%s"`,
-				otherSideKey.Desc(),
-				thisSideKey.Desc(),
-				relation.Type(),
-			)
-			m.Logger.Warn(`Warning: `, err)
+			// Remove invalid relation
+			relations.Remove(relation)
+			errors.Append(fmt.Errorf(`%s not found`, otherSideKey.Desc()))
```
Comment (webrouse) [inline internal/pkg/mapper/relations/load.go:72]: Prehladnejsi vypis chyb, vid testy.

File: internal/pkg/mapper/relations/load.go [hunk 13]
```diff
@@ -27,91 +27,91 @@ func (m *relationsMapper) MapAfterLocalLoad(recipe *model.LocalLoadRecipe) error
...
 		} else {
-			errors.Append(fmt.Errorf(
-				`%s cannot have Relations, referenced from %s, by relation "%s"`,
-				otherSideKey.Desc(),
-				thisSideKey.Desc(),
-				relation.Type(),
-			))
+			// Remove invalid relation
+			relations.Remove(relation)
+			errors.Append(fmt.Errorf(`%s cannot have relation`, otherSideKey.Desc()))
```
Comment (webrouse) [inline internal/pkg/mapper/relations/load.go:84]: Prehladnejsi vypis chyb, vid testy.

File: internal/pkg/mapper/variables/save.go [hunk 14]
```diff
@@ -62,20 +62,8 @@ func (m *variablesMapper) saveVariablesValues(apiObject, internalObject *model.C
 	}
 	relation := relationRaw.(*model.VariablesValuesFromRelation)
 
-	// Check variables and values are from same config
-	variablesConfig := relation.Source.ConfigKey(internalObject.BranchKey())
```
Comment (webrouse) [inline internal/pkg/mapper/variables/save.go:66]: Toto uz nie je potrebne, ako som uz pisal `variablesValuesFor` uz neumoznuje ulozit duplicitne IDs.
Target IDs su iba v relacii `variablesFor`.

File: internal/pkg/model/keys.go [hunk 15]
```diff
@@ -48,24 +48,13 @@ type ConfigKey struct {
 	Id          string `json:"id" validate:"required"`
 }
 
-type ConfigKeySameBranch struct {
```
Comment (webrouse) [inline internal/pkg/model/keys.go:51]: Toto sa velmi neosvedcilo, a skor to cely kod komplikuje -> mazem.

File: internal/pkg/model/keys.go [hunk 16]
```diff
@@ -48,24 +48,13 @@ type ConfigKey struct {
...
-}
-
 type ConfigRowKey struct {
 	BranchId    int    `json:"-" validate:"required"`
 	ComponentId string `json:"-" validate:"required"`
 	ConfigId    string `json:"-" validate:"required"`
 	Id          string `json:"id" validate:"required" `
 }
 
-type ConfigRowKeySameBranch struct {
```
Comment (webrouse) [inline internal/pkg/model/keys.go:63]: Toto sa velmi neosvedcilo, a skor to cely kod komplikuje -> mazem.

File: internal/pkg/model/variables.go [hunk 17]
```diff
@@ -137,18 +157,52 @@ func (t *VariablesValuesForRelation) IsDefinedInApi() bool {
...
+		errors := utils.NewMultiError()
+		errors.Append(fmt.Errorf(`missing relation "%s" in %s`, VariablesForRelType, variablesConfig.Desc()))
+		errors.AppendRaw(fmt.Sprintf(`  - referenced from %s`, relationDefinedOn.Desc()))
+		errors.AppendRaw(fmt.Sprintf(`  - by relation "%s"`, t.Type()))
+		return nil, nil, errors
+	}
+	variablesForRelation := variablesForRaw.(*VariablesForRelation)
+	otherSide := ConfigKey{
+		BranchId:    variablesConfig.BranchId,
+		ComponentId: variablesForRelation.ComponentId,
```
Comment (webrouse) [inline internal/pkg/model/variables.go:188]: `variablesValuesFor` relacia, ziskava target IDs z `variablesFor`, ktora musi tiez existovat.

File: internal/pkg/model/variables.go [hunk 18]
```diff
@@ -179,16 +229,37 @@ func (t *VariablesValuesFromRelation) IsDefinedInApi() bool {
...
+		errors.Append(fmt.Errorf(`missing relation "%s" in %s`, VariablesFromRelType, config.Desc()))
+		errors.AppendRaw(fmt.Sprintf(`  - referenced from %s`, relationDefinedOn.Desc()))
+		errors.AppendRaw(fmt.Sprintf(`  - by relation "%s"`, t.Type()))
+		return nil, nil, errors
+	}
+
+	otherSideKey := ConfigRowKey{
+		BranchId:    config.BranchId,
+		ComponentId: VariablesComponentId,
+		ConfigId:    variablesFromRel.(*VariablesFromRelation).VariablesId,
```
Comment (webrouse) [inline internal/pkg/model/variables.go:252]: A aj opacne, `VariablesValuesFROM` ziskava config id z `VariablesFROM`.

**Review by webrouse — COMMENTED** [review]
V tomto PRs je clean up po implementacii relations a variables:
- pridane chybajuce podmienky a nove chyby
- v `manifest.json` nie su uz duplicitne IDs pri relacii `variablesValuesFor`, vid komentare
- prehladnejsi vypis chyb
- odstranene `ConfigKeySameBranch` a `ConfigRowKeySameBranch` -> jednoduchsi kod.

Nemam to velmi cas delit na mensie PRs, s komentarmi by sa malo dat z toho vyjst, ak nie, tak zavolajme.

**Review by JakubMatejka — APPROVED** [review]
